model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt

```diff
@@ -45,32 +45,36 @@ class VersionMerger(private val storeCache: IDeserializingKeyValueStore, private
         }
     }
 
-    private fun collectLatestNonMerges(version: CLVersion?, visited: MutableSet<String>, result: MutableSet<Long>) {
-        if (version == null) return
-        if (!visited.add(version.getContentHash())) return
-        if (version.isMerge()) {
-            collectLatestNonMerges(version.getMergedVersion1(), visited, result)
-            collectLatestNonMerges(version.getMergedVersion2(), visited, result)
-        } else {
-            result.add(version.id)
-        }
-    }
-
     protected fun mergeHistory(leftVersion: CLVersion, rightVersion: CLVersion): CLVersion {
         if (leftVersion.hash == rightVersion.hash) return leftVersion
-        val commonBase = Companion.commonBaseVersion(leftVersion, rightVersion)
+        val commonBase = commonBaseVersion(leftVersion, rightVersion)
         if (commonBase?.hash == leftVersion.hash) return rightVersion
         if (commonBase?.hash == rightVersion.hash) return leftVersion
 
-        val leftNonMerges = HashSet<Long>().also { collectLatestNonMerges(leftVersion, HashSet(), it) }
-        val rightNonMerges = HashSet<Long>().also { collectLatestNonMerges(rightVersion, HashSet(), it) }
+        val leftVersions = LinearHistory(commonBase?.hash).computeHistoryLazy(leftVersion)
+        if (leftVersions.contains(rightVersion)) {
+            // No merge needed, fast-forward to the left version
+            return leftVersion
+        }
+        val rightVersions = LinearHistory(commonBase?.hash).computeHistoryLazy(rightVersion)
+        if (rightVersions.contains(leftVersion)) {
+            // No merge needed, fast-forward to the right version
+            return rightVersion
+        }
+        val leftNonMerges = leftVersions.filterNot { it.isMerge() }
+        val rightNonMerges = rightVersions.filterNot { it.isMerge() }
+        // We do not have to compare sets.
+        // If the left and right versions have the same versions (excluding merges) in the same order,
+        // then their contents will be the same.
+        // If the left and right versions have different versions (excluding merges) or a different order,
+        // then their content might be different, and we need to decide on a new order in a new merge.
```

Comment:
I think this `if` branch is important to avoid the described endless creation of merge commits. Even if the order of both sides is different I would still expect the new order in `versionsToApply` to be equal to one of the sides, in most cases. We should check that and then return the left/right side instead of creating a new merge commit.

Reply:
Done.